hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/procedures/ShowFileHistoryProcedureUtils.scala

```diff
@@ -0,0 +1,415 @@
+      totalUpsertTimeMs = -1,
+      totalCreateTimeMs = -1,
+      prevBaseFile = None,
+      columnStatsAvailable = false
+    )
+  }
+
+  def checkForDeletionsAndReplacements(metaClient: HoodieTableMetaClient,
+                                       fileGroupId: String,
+                                       targetPartition: Option[String],
+                                       showArchived: Boolean): (Map[String, DeletionInfo], Map[String, ReplacementInfo]) = {
+
+    val deletions = scala.collection.mutable.Map[String, DeletionInfo]()
+    val replacements = scala.collection.mutable.Map[String, ReplacementInfo]()
+
+    checkDeletionsAndReplacementsInTimeline(metaClient.getActiveTimeline, fileGroupId, targetPartition, "ACTIVE", deletions, replacements)
+
+    if (showArchived) {
+      try {
+        val archivedTimeline = metaClient.getArchivedTimeline.reload()
+        archivedTimeline.loadCompletedInstantDetailsInMemory()
+        checkDeletionsAndReplacementsInTimeline(archivedTimeline, fileGroupId, targetPartition, "ARCHIVED", deletions, replacements)
+      } catch {
+        case e: Exception =>
+          log.warn(s"Failed to check deletions in archived timeline: ${e.getMessage}")
+      }
+    }
+
+    (deletions.toMap, replacements.toMap)
+  }
+
+  def checkDeletionsAndReplacementsInTimeline(timeline: HoodieTimeline,
+                                              fileGroupId: String,
+                                              targetPartition: Option[String],
+                                              timelineType: String,
+                                              deletions: scala.collection.mutable.Map[String, DeletionInfo],
+                                              replacements: scala.collection.mutable.Map[String, ReplacementInfo]): Unit = {
+
+    val cleanInstants = timeline.getCleanerTimeline.getInstants.iterator().asScala
+    for (instant <- cleanInstants) {
+      try {
+        if (instant.getState == HoodieInstant.State.COMPLETED) {
+          val cleanMetadata = timeline.readCleanMetadata(instant)
+          val partitionsToCheck = targetPartition.map(Set(_)).getOrElse(cleanMetadata.getPartitionMetadata.keySet().asScala.toSet)
+
+          for {
+            partitionPath <- partitionsToCheck
+            partitionMetadata <- Option(cleanMetadata.getPartitionMetadata.get(partitionPath))
+            deletedFile <- partitionMetadata.getSuccessDeleteFiles.asScala
+            if matchesDeletedFileGroup(deletedFile, fileGroupId)
+          } {
+            val deletedFileName = extractActualFileName(deletedFile)
+            deletions(deletedFileName) = DeletionInfo(HoodieTimeline.CLEAN_ACTION, instant.requestedTime, timelineType)
+          }
+        }
+      } catch {
+        case e: Exception =>
+          log.warn(s"Failed to process clean instant ${instant.requestedTime}: ${e.getMessage}")
+      }
+    }
+
+    val rollbackInstants = timeline.getRollbackTimeline.getInstants.iterator().asScala
```

Comment:
not required right away. can you file a follow up to add `restore` instants processing as well as part of these procedures